In [6]:
import pandas as pd
import numpy as np
import seaborn as sns

# 1. Ingesta de data y análisis de esta

In [28]:
df_2009 = pd.read_csv("2009_1perc.csv").drop('Unnamed: 0', axis=1)
df_2010 = pd.read_csv("2010_1perc.csv").drop('Unnamed: 0', axis=1)

In [34]:
df_2009.head(3)

year  pct  ser_num  datestop  timestop recstat inout trhsloc  perobs  \
0  2009   41     1779   4032009       130       A     O       P     1.0   
1  2009  108     5805  10292009      1050       A     O       P     3.0   
2  2009   43     8340  10062009      1450       1     O       P     1.0   

  crimsusp  ...  zip addrpct sector beat post   xcoord   ycoord dettypcm  \
0      CPW  ...           41      G    7       1013067  0238633       CM   
1     BURG  ...          108      J            1012043  0212157       CM   
2     MISD  ...           43      E            1017599  0240200       CM   

   linecm  detailcm  
0       1        20  
1       1        14  
2       1        20  

[3 rows x 111 columns]

In [35]:
df_2010.head(3)

year  pct  ser_num  datestop  timestop recstat inout trhsloc  perobs  \
0  2010    6      487   1192010      1420       1     O       P       2   
1  2010   42     7138   9042010       225       1     I       T      10   
2  2010   33     2628   6302010      1240       1     I       T       5   

  crimsusp  ...  zip addrpct sector beat post   xcoord  ycoord dettypcm  \
0      FEL  ...            6      G        38   983543  208767       CM   
1   FELONY  ...           42      E            1014172  241737       CM   
2  ROBBERY  ...           33      B    6       1000916  245710       CM   

   linecm  detailcm  
0       1        85  
1       1        45  
2       1        85  

[3 rows x 111 columns]

In [40]:
print(f'datos 2009: {df_2009.shape[1]} columnas y {df_2009.shape[0]} datos')
print(f'datos 2010: {df_2010.shape[1]} columnas y {df_2010.shape[0]} datos')


datos 2009: 111 columnas y 5812 datos
datos 2010: 111 columnas y 6013 datos


Los datos de ambos DataFrames tienen 111 columnas, que lo más probable que no todas sean de utilidad, además algunas columnas como 'beat' o 'post' tienen registros con datos faltantes, en formato de espacios. Veremos más a profundidad estos datos faltantes.

In [43]:
df_2009.replace(' ', np.nan, inplace=True)
df_2010.replace(' ', np.nan, inplace=True)

In [51]:
df_2009.isna().sum().sort_values(ascending=False)[:23]

aptnum      5812
othfeatr    5812
rescode     5812
premtype    5812
zip         5812
state       5812
officrid    5727
arstoffn    5464
sumoffen    5456
post        5268
offverb     4559
offshld     4205
beat        3622
addrnum     3300
stname      3225
premname     591
trhsloc      205
xcoord       168
ycoord       168
sector       116
stinter       40
crossst       40
cs_other       0
dtype: int64

Como podemos ver en la lista de arriba, se aprecia que hay 22 columnas con valores perdidos, pero lo interesante es en cuanto a la cantidad de datos perdidos, se puede apreciar que desde "aptnum" a "stname" pasan de 5.812 a 3.225 datos perdidos respectivamente, despues baja drásticamente a 591 con "premname" y siguen siendo menos los datos perdidos.

Lo interesante es que estos datos perdidos hay que tratarlos, pero contamos con una función auxiliar que hace este trabajo, de imputar valores en algunas columnas, según algunos criterios ahí definidos, además de reglas como por ejemplo que la edad de la persona esté entre 18 y 100 años, de modo que si viola esta regla, le asigna valos np.nan, eliminando estos registros posteriormente (No deberían poder controlar gente menor de edad y si tiene más de 100 años es muy probable que sea un error de tipeo).